src/main/java/com/minecolonies/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -828,19 +828,22 @@ private static ItemStack handleKeepX(@NotNull Map<ItemStorage, Integer> keptX,
      */
     private static boolean keptEnough(@NotNull Map<ItemStorage, Integer> kept, @NotNull Map<ItemStorage, Integer> keep, @NotNull ItemStack stack)
     {
-        Set<ItemStorage> tempKeep = new HashSet<>(keep.keySet());
-        for(ItemStorage tempStack: tempKeep)
+        final ArrayList<Map.Entry<ItemStorage, Integer>> tempKeep = new ArrayList<>(keep.entrySet());
+        for(Map.Entry<ItemStorage, Integer> tempEntry: tempKeep)
```

Comment:
Shouldn't make the list if we are using entrySet's , we want to just use the iterator, so the foreach should use keep.entrySet. But kind of besides the point now, because I think we need to remove the loop.